rules/private/phases/phase_bootstrap_compile.bzl

```diff
@@ -18,54 +18,118 @@ def phase_bootstrap_compile(ctx, g):
 
     inputs = depset(
         ctx.files.srcs,
-        transitive = [ctx.attr._jdk[java_common.JavaRuntimeInfo].files, g.classpaths.compile, g.classpaths.compiler],
+        transitive = [
+            ctx.attr._jdk[java_common.JavaRuntimeInfo].files,
+            g.classpaths.compile,
+            g.classpaths.compiler,
+        ],
     )
 
     tmp = ctx.actions.declare_directory("{}/tmp/classes".format(ctx.label.name))
 
-    compiler_classpath = ":".join([f.path for f in g.classpaths.compiler.to_list()])
-    compile_classpath = ":".join([f.path for f in g.classpaths.compile.to_list()])
-    srcs = " ".join([f.path for f in g.classpaths.srcs])
-
     scala_configuration = g.javainfo.scala_info.scala_configuration
 
     main_class = "scala.tools.nsc.Main"
     if int(scala_configuration.version[0]) >= 3:
         main_class = "dotty.tools.dotc.Main"
 
+    compiler_classpath = g.classpaths.compiler.to_list()
+    compile_classpath = g.classpaths.compile.to_list()
+
+    args = ctx.actions.args()
+    args.add("--java", ctx.attr._jdk[java_common.JavaRuntimeInfo].java_executable_exec_path)
+    args.add("--main_class", main_class)
+    args.add("--jar_creator", ctx.executable._jar_creator)
+    args.add("--output_jar", g.classpaths.jar)
+
+    if compiler_classpath:
+        args.add_joined("--compiler_classpath", compiler_classpath, join_with = ":")
+    else:
+        fail("Compiler classpath  missing for bootstrap compiler")
+
+    if compile_classpath:
+        args.add_joined("--compile_classpath", compile_classpath, join_with = ":")
+
+    args.add_all("--tmp", [tmp], expand_directories = False)
+
+    if scala_configuration.global_scalacopts:
+        args.add_joined("--global_scalacopts", scala_configuration.global_scalacopts, join_with = " ")
+
+    if ctx.attr.scalacopts:
+        args.add_joined("--scalacopts", ctx.attr.scalacopts, join_with = " ")
+
+    if g.classpaths.srcs:
+        args.add_joined("--srcs", g.classpaths.srcs, join_with = " ")
+    else:
+        fail("Empty srcs list passed to bootstrap compiler")
```

Comment:
Until now, our rules have supported an empty source list. Shouldn't it continue to be valid? Or does the Scala compiler not support this?

Reply:
I'll remove the `fail` here. I think whether it will work depends on what the other arguments are.